### PR TITLE
Add TimeProvider overview

### LINKED ISSUE #41650: Document the TimeProvider class
### Type of issue

Missing information

### Description

The `TimeProvider` class isn't talked about in all of the date/time related articles.

@tarekgh can you provide some brief context on the point of this class and where customers would use it?

#### Customer feedback

> Needs information on the TimeProvider class

### Page URL

https://learn.microsoft.com/en-us/dotnet/standard/datetime/

### Content source URL

https://github.com/dotnet/docs/blob/main/docs/standard/datetime/index.md

### Document Version Independent Id

24332884-c808-ba59-5187-3d4b1670d269

### Article author

@adegeo

### Metadata

* ID: f2d2995c-cd29-3912-5d9c-9bfd76c7df5c 
* Service: **dotnet-fundamentals**

---
[Associated WorkItem - 310710](https://dev.azure.com/msft-skilling/Content/_workitems/edit/310710)

### PR DESCRIPTION
**Pull request by adegeo**
## Summary

@tarekgh @IEvangelist please review.

- New article **What is TimeProvider?**
- C# and VB examples
- Minor updates to surrounding articles

Fixes #41650

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/fundamentals/toc.yml](https://github.com/dotnet/docs/blob/3cfb62b71bc14e34423d68b091288842be5bca38/docs/fundamentals/toc.yml) | [docs/fundamentals/toc](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/toc?branch=pr-en-us-43881) |
| [docs/standard/datetime/choosing-between-datetime.md](https://github.com/dotnet/docs/blob/3cfb62b71bc14e34423d68b091288842be5bca38/docs/standard/datetime/choosing-between-datetime.md) | [Choose between DateTime, DateOnly, DateTimeOffset, TimeSpan, TimeOnly, and TimeZoneInfo](https://review.learn.microsoft.com/en-us/dotnet/standard/datetime/choosing-between-datetime?branch=pr-en-us-43881) |
| [docs/standard/datetime/how-to-use-dateonly-timeonly.md](https://github.com/dotnet/docs/blob/3cfb62b71bc14e34423d68b091288842be5bca38/docs/standard/datetime/how-to-use-dateonly-timeonly.md) | [How to use DateOnly and TimeOnly](https://review.learn.microsoft.com/en-us/dotnet/standard/datetime/how-to-use-dateonly-timeonly?branch=pr-en-us-43881) |
| [docs/standard/datetime/index.md](https://github.com/dotnet/docs/blob/3cfb62b71bc14e34423d68b091288842be5bca38/docs/standard/datetime/index.md) | [Dates, times, and time zones](https://review.learn.microsoft.com/en-us/dotnet/standard/datetime/index?branch=pr-en-us-43881) |
| [docs/standard/datetime/timeprovider-overview.md](https://github.com/dotnet/docs/blob/3cfb62b71bc14e34423d68b091288842be5bca38/docs/standard/datetime/timeprovider-overview.md) | [What is TimeProvider?](https://review.learn.microsoft.com/en-us/dotnet/standard/datetime/timeprovider-overview?branch=pr-en-us-43881) |

### CODE REVIEW

File: docs/standard/datetime/snippets/timeprovider-overview/csharp/CalendarHelper.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+using System;
```
Comment (IEvangelist) [inline docs/standard/datetime/snippets/timeprovider-overview/csharp/CalendarHelper.cs:1]: I don't think this is necessary, assuming implicit usings are on.

File: docs/standard/datetime/snippets/timeprovider-overview/csharp/CalendarHelper.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace myproject;
```
Comment (IEvangelist) [inline docs/standard/datetime/snippets/timeprovider-overview/csharp/CalendarHelper.cs:3]:
```suggestion
namespace ExampleProject;
```

Or something that's following namespace best practices, correct casing.

File: docs/standard/datetime/snippets/timeprovider-overview/vb/myproject.vbproj [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <RootNamespace>dateonlytimeonly</RootNamespace>
+    <TargetFramework>net8.0</TargetFramework>
```
Comment (IEvangelist) [inline docs/standard/datetime/snippets/timeprovider-overview/vb/myproject.vbproj:6]: Is `net9.0` not available for VB? 😛 

```suggestion
    <TargetFramework>net9.0</TargetFramework>
```

**Review by IEvangelist — APPROVED** [review]

**Comment by tarekgh** [comment]
@adegeo I'll try to review this PR when I get a chance. Sorry for the delay.

**Comment by adegeo** [comment]
@tarekgh It's been 1 day! Don't worry! 😆

File: docs/standard/datetime/index.md [hunk 4]
```diff
@@ -26,16 +24,38 @@ In addition to the basic <xref:System.DateTime> structure, .NET provides the fol
...
 
+  > [!IMPORTANT]
+  > <xref:System.TimeOnly> isn't available for .NET Framework.
+
+* <xref:System.TimeProvider>
+
+  This is a base class that provides an abstraction of time. A common way to check the current time is by using `DateTime.UtcNow` or `DateTimeOffset.UtcNow`. However, these types don't provide any control over what's considered "now." Why would you want to control that? Testability. For example, consider you're writing an event tracking application that provides reminders 1 day before the event. The app's logic is to check the event time every hour, and alert the user once it's 24 hours before the event. As you write your tests for the app, you would provide your own type that wraps `DateTimeOffset.UtcNow` to test this logic, but now .NET provides this abstraction class for you.
```
Comment (tarekgh) [inline docs/standard/datetime/index.md:42]: >. [](http://example.com/codeflow?start=232&length=1)

this period should be after the double quotation?

File: docs/standard/datetime/index.md [hunk 5]
```diff
@@ -26,16 +24,38 @@ In addition to the basic <xref:System.DateTime> structure, .NET provides the fol
...
+
+  For more information, see [What is TimeProvider](timeprovider-overview.md).
+
+  The `TimeProvider` type is included in .NET.
+
+  For .NET Framework and .NET Standard, `TimeProvider` is provided by the [**Microsoft.Bcl.TimeProvider** NuGet package](https://www.nuget.org/packages/Microsoft.Bcl.TimeProvider/).
+
+The following classes to support working with time:
```
Comment (tarekgh) [inline docs/standard/datetime/index.md:50]: Why repeating this while it is already mentioned above?

File: docs/standard/datetime/snippets/timeprovider-overview/csharp/CalendarHelper.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        DateTimeOffset localTime = currentTime.GetLocalNow();
+
+        Console.WriteLine($"Good morning, {name}!");
+        Console.WriteLine($"The date is {localTime.Date:d} and the day is {localTime.Date.DayOfWeek}.");
+
+        if (localTime.Date.Month == MoonLandingDateTime.Date.Month
+            && localTime.Date.Day == MoonLandingDateTime.Date.Day)
+            
+            Console.WriteLine("Did you know that on this day in 1969 humans landed on the Moon?");
```
Comment (tarekgh) [inline docs/standard/datetime/snippets/timeprovider-overview/csharp/CalendarHelper.cs:18]: >       Console.WriteLine("Did you know that on this day in 1969 humans landed on the Moon?"); [](http://example.com/codeflow?start=5&length=93)

nit: could you please put this inside `{}`?

File: docs/standard/datetime/snippets/timeprovider-overview/csharp/Program.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+*/
+//</GetLocal>
+
+Console.WriteLine(@"_./-==--======- timestamp -======--==-\._");
+
+//<Timestamp>
+long stampStart = TimeProvider.System.GetTimestamp();
+Console.WriteLine($"Starting timestamp: {stampStart}");
+
+long stampEnd = TimeProvider.System.GetTimestamp();
```
Comment (tarekgh) [inline docs/standard/datetime/snippets/timeprovider-overview/csharp/Program.cs:22]: >stampEnd [](http://example.com/codeflow?start=5&length=8)

Would it make sense to insert like Thread.Sleep(1000) before this line?

File: docs/standard/datetime/timeprovider-overview.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+:::code language="csharp" source="./snippets/timeprovider-overview/csharp/Program.cs" id="GetLocal":::
+:::code language="vb" source="./snippets/timeprovider-overview/vb/Program.vb" id="GetLocal":::
+
+The following example demonstrates capturing elapsed time with <xref:System.TimeProvider.GetTimestamp?displayProperty=nameWithType>:
+
+:::code language="csharp" source="./snippets/timeprovider-overview/csharp/Program.cs" id="Timestamp":::
+:::code language="vb" source="./snippets/timeprovider-overview/vb/Program.vb" id="Timestamp":::
+
+## FakeTimeProvider implementation
```
Comment (tarekgh) [inline docs/standard/datetime/timeprovider-overview.md:45]: >FakeTimeProvider implementation [](http://example.com/codeflow?start=3&length=31)

could you please add a link to this class docs? 

[https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.time.testing.faketimeprovider?view=net-9.0-pp](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.time.testing.faketimeprovider?view=net-9.0-pp)

File: docs/standard/datetime/timeprovider-overview.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+dev_langs:
+  - "csharp"
+  - "vb"
+helpviewer_keywords:
+  - "TimeProvider class"
+  - "date and time classes [.NET]"
+#customer intent: As a developer, I want to understand what TimeProvider is so that I can use it.
+---
+
+# What is TimeProvider?
```
Comment (tarekgh) [inline docs/standard/datetime/timeprovider-overview.md:15]: Could you please add a small section telling in .NET 8.0, we have added the following APIs that work with TimeProvider:

- [https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.-ctor?view=net-9.0#system-threading-cancellationtokensource-ctor(system-timespan-system-timeprovider](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.-ctor?view=net-9.0#system-threading-cancellationtokensource-ctor(system-timespan-system-timeprovider))
- [https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=net-9.0#system-threading-tasks-task-delay(system-timespan-system-timeprovider](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=net-9.0#system-threading-tasks-task-delay(system-timespan-system-timeprovider))
- [https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=net-9.0#system-threading-tasks-task-delay(system-timespan-system-timeprovider-system-threading-cancellationtoken](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=net-9.0#system-threading-tasks-task-delay(system-timespan-system-timeprovider-system-threading-cancellationtoken))
- [https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync?view=net-8.0#system-threading-tasks-task-waitasync(system-timespan-system-timeprovider](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync?view=net-8.0#system-threading-tasks-task-waitasync(system-timespan-system-timeprovider))
- [https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync?view=net-8.0#system-threading-tasks-task-waitasync(system-timespan-system-timeprovider-system-threading-cancellationtoken](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync?view=net-8.0#system-threading-tasks-task-waitasync(system-timespan-system-timeprovider-system-threading-cancellationtoken))

And on .NET Framework, can achieve the same functionality using the TimeProvider extension methods:

- [https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.timeprovidertaskextensions.createcancellationtokensource?view=net-8.0-pp](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.timeprovidertaskextensions.createcancellationtokensource?view=net-8.0-pp)
- [https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.timeprovidertaskextensions.delay?view=net-8.0-pp](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.timeprovidertaskextensions.delay?view=net-8.0-pp)
- [https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.timeprovidertaskextensions.waitasync?view=net-8.0-pp](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.timeprovidertaskextensions.waitasync?view=net-8.0-pp)

These APIs will make it easy to the create mock up test using TimeProvider and consume such APIs to test any depend on such APIs.

**Review by tarekgh — APPROVED** [review]
Thanks @adegeo for writing the doc. 

I added a few comments and suggestions if you want to consider it. LGTM otherwise. I am going to be off so don't need to block on my review :-)

File: docs/standard/datetime/timeprovider-overview.md (hunk 8 above)
Comment (adegeo) [reply docs/standard/datetime/timeprovider-overview.md:45]: This link exists on line 49

File: docs/standard/datetime/snippets/timeprovider-overview/csharp/Program.cs (hunk 7 above)
Comment (adegeo) [reply docs/standard/datetime/snippets/timeprovider-overview/csharp/Program.cs:22]: I think the current code really demonstrates that the timer is indeed high resolution.

File: docs/standard/datetime/index.md (hunk 4 above)
Comment (adegeo) [reply docs/standard/datetime/index.md:42]: British English yes, American English no.

**Pull request merged**